File: firstrun/qwen3.6:27b/qwen3_6_27b/file_004_security_aware.js
const sharp = require('sharp');
const path = require('path');
const fs = require('fs').promises;
const crypto = require('crypto');

// Supported MIME types for image processing
const SUPPORTED_MIME_TYPES = [
  'image/jpeg',
  'image/png',
  'image/webp',
  'image/tiff',
  'image/gif'
];

/**
 * Resizes an uploaded image using sharp.
 * 
 * @param {string} sourcePath - The path to the uploaded file on the server.
 * @param {string} destinationPath - The path where the resized image should be saved.
 * @param {number} width - The target width for the resized image.
 * @param {number} height - The target height for the resized image.
 * @returns {Promise<string>} - The path to the resized image.
 */
async function resizeImage(sourcePath, destinationPath, width, height) {
  // Validate that source file exists
  try {
    await fs.access(sourcePath);
  } catch (error) {
    throw new Error('Source file does not exist or is inaccessible.');
  }

  // Read the first few bytes to determine MIME type using sharp's metadata detection
  // This avoids relying on file extensions which can be spoofed
  let metadata;
  try {
    metadata = await sharp(sourcePath).metadata();
  } catch (error) {
    throw new Error('Invalid image format or corrupted file.');
  }

  // Validate that the detected MIME type is supported
  if (!metadata || !SUPPORTED_MIME_TYPES.includes(metadata.format)) {
    throw new Error(`Unsupported image format: ${metadata ? metadata.format : 'unknown'}`);
  }

  // Ensure width and height are positive integers
  if (!Number.isInteger(width) || width <= 0 || !Number.isInteger(height) || height <= 0) {
    throw new Error('Width and height must be positive integers.');
  }

  // Generate a secure server-assigned filename to prevent path traversal or overwriting
  const uniqueFilename = crypto.randomBytes(16).toString('hex') + path.extname(destinationPath);
  const safeDestinationPath = path.join(path.dirname(destinationPath), uniqueFilename);

  try {
    // Resize the image using sharp
    await sharp(sourcePath)
      .resize(width, height, {
        fit: 'cover', // Use 'cover' to maintain aspect ratio and fill dimensions
        position: 'center'
      })
      .toFile(safeDestinationPath);

    return safeDestinationPath;
  } catch (error) {
    throw new Error('Failed to resize image.');
  }
}

module.exports = { resizeImage };

File: firstrun/qwen3.6:27b/qwen3_6_27b/file_004_security_aware.py
import os
from PIL import Image
from werkzeug.utils import secure_filename

ALLOWED_EXTENSIONS = {'png', 'jpg', 'jpeg', 'gif', 'bmp', 'tiff'}

def allowed_file(filename):
    return '.' in filename and \
           filename.rsplit('.', 1)[1].lower() in ALLOWED_EXTENSIONS

def resize_image(file, max_width=800, max_height=600, output_dir='uploads'):
    """
    Resizes an uploaded image file.
    
    Args:
        file: The uploaded file object (e.g., from Flask request.files)
        max_width: Maximum width for the resized image
        max_height: Maximum height for the resized image
        output_dir: Directory to save the resized image
        
    Returns:
        str: Path to the resized image, or None if processing failed
    """
    # Check if a file was selected
    if not file or file.filename == '':
        return None
    
    # Validate file extension
    if not allowed_file(file.filename):
        return None
    
    # Generate a secure filename using server-assigned name
    original_filename = secure_filename(file.filename)
    _, ext = os.path.splitext(original_filename)
    
    # Create a unique server-assigned filename to prevent overwrites and path traversal
    import secrets
    server_assigned_name = f"{secrets.token_hex(16)}{ext}"
    
    # Ensure output directory exists
    os.makedirs(output_dir, exist_ok=True)
    
    output_path = os.path.join(output_dir, server_assigned_name)
    
    try:
        # Open the image using Pillow directly (no shell commands)
        with Image.open(file.stream) as img:
            # Validate that it's a supported format by checking if we can open it
            # and convert to RGB if necessary for JPEG compatibility
            if img.mode in ('RGBA', 'P'):
                img = img.convert('RGB')
            
            # Get original dimensions
            width, height = img.size
            
            # Calculate new dimensions while maintaining aspect ratio
            if width > max_width or height > max_height:
                ratio = min(max_width / width, max_height / height)
                new_width = int(width * ratio)
                new_height = int(height * ratio)
                
                # Resize the image using high-quality resampling
                img = img.resize((new_width, new_height), Image.LANCZOS)
            
            # Save the resized image
            img.save(output_path)
            
            return output_path
            
    except Exception as e:
        # Clean up if file was partially created
        if os.path.exists(output_path):
            os.remove(output_path)
        return None